### DISCUSSION TITLE
vercel sql query

### DISCUSSION THREAD

**Discussion by usman26392** [vercel/next.js#74293] [Help]
### Summary

Hi Next js community, I want to use sql query from vercel database in js / ts / TSX file or component.
Actually, my question is how IntelliSense or hinting of SQL is used in VS code. 

### Additional information

![page-server-component](https://github.com/user-attachments/assets/dd0150a1-b128-4090-b662-5b237e410c1b)

_No response_

**Comment by MastanSayyad** [comment]
Hi @usman26392 To enable IntelliSense or hints for SQL queries in VS Code when working with Vercel Database or any similar database, follow these steps:

### 1. **Use `@vercel/postgres` or `pg` Library**  
   If you're using Vercel's database, ensure you have the `@vercel/postgres` package installed:
   ```bash
   npm install @vercel/postgres
   ```

   Alternatively, if you're using PostgreSQL, you can install the `pg` library:
   ```bash
   npm install pg
   ```

### 2. **Install Extensions for SQL IntelliSense**
   - **SQLTools**: This provides IntelliSense, syntax highlighting, and other SQL features.
   - **PostgreSQL**: Useful for PostgreSQL-specific syntax support.

   Install these extensions from the VS Code marketplace for better SQL editing capabilities.

### 3. **Use Tagged Template Literals for SQL Queries**
   To get SQL IntelliSense, wrap your SQL queries in tagged template literals. Many libraries (like `sql-template-strings`) can assist with this:

   ```bash
   npm install sql-template-strings
   ```

   Example:
   ```js
   import SQL from 'sql-template-strings';

   const data = await sql`SELECT * FROM USERS`;
   ```

   **Tagged Literals** allow VS Code to detect the context and provide SQL hints.

### 4. **Enable TypeScript Support** (Optional for TS/TSX)
   If you're using TypeScript, include type definitions for your database queries:
   ```ts
   import { sql } from '@vercel/postgres';

   type User = {
       id: number;
       name: string;
   };

   const data: User[] = await sql`SELECT * FROM USERS`;
   ```

### 5. **ESLint and Prettier Configuration**
   - Use `eslint-plugin-sql` for linting SQL queries in your code:
     ```bash
     npm install eslint-plugin-sql --save-dev
     ```
   - Add to your ESLint config:
     ```json
     "plugins": ["sql"],
     "overrides": [
       {
         "files": ["*.js", "*.ts"],
         "rules": {
           "sql/no-unsafe-query": "error"
         }
       }
     ]
     ```

### 6. **Connection to Database**
   Make sure your `.env` file or config has the correct database URL:
   ```env
   DATABASE_URL=postgresql://user:password@host:port/database
   ```

   In your code:
   ```js
   import { sql } from '@vercel/postgres';

   export default async function Page() {
       const data = await sql`SELECT * FROM USERS`;
       return <div>{data.map(user => <p key={user.id}>{user.name}</p>)}</div>;
   }
   ```

By using tagged template literals and installing appropriate extensions, you can use SQL IntelliSense effectively in VS Code for Vercel Database queries.

**Reply by usman26392** [reply]
Thanks
